### ISSUE TITLE
Poor assumption about FS permissions

### ISSUE THREAD

**Issue by mlucool** [jupyter-lsp/jupyterlab-lsp#924]
## Description

LSP assumes that you have write access to root_dir [here](https://github.com/jupyter-lsp/jupyterlab-lsp/blob/b7690aae90a7d85cc422c8f934ed5bf992a5f902/python_packages/jupyter_lsp/jupyter_lsp/serverextension.py#L46). This tries to create a `.virtual_documents` directory.

## Reproduce

Set root_dir to `/` on most OS. The logs server logs will now include an error like:
```
ERROR 10:55:52 Exception in callback functools.partial(<bound method IOLoop._discard_future_result of <tornado.platform.asyncio.AsyncIOMainLoop object at 0x7f883de2e140>>, <Task finished name='Task-1' coro=<initialize() done, defined at /path/to/python/lib64/python3.10/site-packages/jupyter_lsp/serverextension.py:13> exception=PermissionError(13, 'Permission denied')>)
    Traceback (most recent call last):
      File "/path/to/python/lib64/python3.10/site-packages/tornado/ioloop.py", line 740, in _run_callback
        ret = callback()
      File "/path/to/python/lib64/python3.10/site-packages/tornado/ioloop.py", line 764, in _discard_future_result
        future.result()
      File "/path/to/python/lib64/python3.10/site-packages/jupyter_lsp/serverextension.py", line 24, in initialize
        setup_shadow_filesystem(virtual_documents_uri=virtual_documents_uri)
      File "/path/to/python/lib64/python3.10/site-packages/jupyter_lsp/virtual_documents_shadow.py", line 114, in setup_shadow_filesystem
        shadow_filesystem.mkdir(parents=True, exist_ok=True)
      File "/opt/python/python-3.10/lib64/python3.10/pathlib.py", line 1173, in mkdir
        self._accessor.mkdir(self, mode)
    PermissionError: [Errno 13] Permission denied: '/.virtual_documents'
```

## Expected behavior

A better URI is chosen (maybe in the settings or data dir for jupyter) and/or this is configurable. 

## Context

- Operating System and version: Linux
- JupyterLab version: 3.6
- Language server and version: 4.0.1

**Comment by krassowski** [comment]
The decision to locate `.virtual_documents` directory in presumed project directory was motivated by limitations of specific language servers; in particular (at the time) python-language-server relied on Python linters some of which assumed that importable packages can be found by traversing the path of current document upwards to the root and trying each directory; if the virtual document is in a directory which has no common path prefix with location of your importable modules then those would not be found and linter will incorrectly show an error on import line.

Three observations:
- the assumption was that the root directory = project directory. This is clearly not the case when setting root directory to `/`.
- the situation evolved since; we have much better configuration options available in `python-lsp-server` (and others) and should be able fix any remaining upstream assumptions about path traversal if needed
- given the above two, we could try to use a temporary directory instead (initially as opt-in and by default in next major version)

> A better URI is chosen (maybe in the settings or data dir for jupyter)

data/config/runtime as provided by `jupyter --paths` can be shared between server instances and this could lead to conflicts (or require a way of generating a subfolder with UUID and managing it).

> or this is configurable

We already have `LanguageServerManager.virtual_documents_dir` traitlet and `JP_LSP_VIRTUAL_DIR` environment variable. Does it address the need here?

**Comment by krassowski** [comment]
> > assumed that importable packages can be found by traversing the path of current document upwards to the root and trying each directory
> 
> In the case where I start a server at `/foo/bar` and then open a notebook at `/foo/bar/baz/qux` and import from a neighbor module `/foo/bar/baz/my_module` wouldn't that get skipped with this assumption?

That was a simplification, what I should have said is that each of the paths was considered for adding to Python interpreter's `sys.path`.

<details>

For example, `python-lsp-server` currently checks for _source_ roots on the way between a file (e.g. `/foo/bar/baz/qux/notebook.ipynb`) and root path (e.g. `/foo/bar`) where _source_ root is expected to have either `setup.py` or `pyproject.toml` (and falls back to root path, this is `/foo/bar`).

```python
def source_roots(self, document_path):
    """Return the source roots for the given document."""
    files = _utils.find_parents(self._root_path, document_path, ['setup.py', 'pyproject.toml']) or []
    return list({os.path.dirname(project_file) for project_file in files}) or [self._root_path]
```

Also, sometimes it just [adds the parent path of the document itself](https://github.com/python-lsp/python-lsp-server/blob/f36fb785ec914c2787fa0c55751088866886f3a3/pylsp/workspace.py#L405-L407):

```python
    # Extend sys_path with document's path if requested
    if use_document_path:
        sys_path += [os.path.normpath(os.path.dirname(self.path))]
```

or just uses the [`root_path` directly](https://github.com/python-lsp/python-lsp-server/blob/f36fb785ec914c2787fa0c55751088866886f3a3/pylsp/workspace.py#L78).

</details>

> Either way, it seems `preferred-dir or root_dir` is a better default than `root_dir`?

Yes, but this would be a bit more involved than changing the default. We may need to rewrite the code to ensure that the contents of `.virtual_documents` are reflecting the full path relative to `root_dir` (rather than relative to location of virtual documents directory).

<details>

Currently, having the `.virtual_documents`  in root dir we can easily convert the paths in server replies to point to the underlying file. For example with root in `/foo/bar/`, if user executes "jump to command" and server responds with `/foo/bar/.virtual_documents/baz/a.py` we know that we should open `/foo/bar/baz/a.py`; nothing changes under your proposal for `/foo/bar/baz/a.py` if the root is `/` and `preferred_dir` is `/foo/bar`, however as soon as user tries to open a file outside of `preferred_dir`, e.g. `/foo/y.py` it would be stored as `/foo/bar/.virtual_documents/y.py` which means we lost 1-1 mapping; hence the need to rewrite `.virtual_documents` handling with new paths being:
- `/foo/bar/.virtual_documents/foo/y.py` for `/foo/y.py`, and
- `/foo/bar/.virtual_documents/foo/bar/baz/a.py`  for `/foo/bar/baz/a.py`

</details>

This could lead to problems on some operating systems, e.g. prior to Windows 10 there is a hard limit of 260 characters on file path length for some API calls; under the rewrite `/a/b/c/.virtual_documents/x.py` would become `/a/b/c/.virtual_documents/a/b/c/x.py` so if `a/b/c` is 150 characters long we would hit the OS limit.

The bigger picture is that we would like to remove the virtual documents altogether as those are a workaround which is not needed for well-behaving servers (https://github.com/jupyter-lsp/jupyterlab-lsp/issues/537). As a first step we could disable use of `.virtual_documents` for specific servers as proposed in #537, then make it opt-in for servers which require it, and finally drop it altogether in future.

**Comment by krassowski** [comment]
To clarify, I believe that the setting the override for `.virtual_documents` using `JP_LSP_VIRTUAL_DIR` or traitlet can break some functionality in the same way as changing the default (as discussed above), unless the override is only used to change the name of the directory (and not its location).

**Comment by mlucool** [comment]
> workaround which is not needed for well-behaving servers (https://github.com/jupyter-lsp/jupyterlab-lsp/issues/537).

Which python servers are considered well behaved? Is pyright?

**Comment by krassowski** [comment]
Yes, as far as I am aware recent versions of pyright do not attempt to access or modify files directly on disk, so it is well behaved in that sense.

It was not always the case, see https://github.com/microsoft/pyright/issues/2181#issuecomment-897793486 which was fixed in 1.1.162 as per https://github.com/microsoft/pyright/issues/2181#issuecomment-898738895.

**Comment by nishikantparmariam** [comment]
https://github.com/jupyter-lsp/jupyterlab-lsp/pull/930 seems to fix things here after disabling `requires_documents_on_disk` for other servers. 

Though there is minor bug - The permission error above can occur on each key press in a notebook or on save when a server with `requires_documents_on_disk=False` is chosen. 

Details:
https://github.com/jupyter-lsp/jupyterlab-lsp/blob/b57638376714d2e48b4a8389570f983d0e3c1932/python_packages/jupyter_lsp/jupyter_lsp/virtual_documents_shadow.py#L146-L154

Initially `initialized` is set to `False`, but if line `149` raises permission error then line `154` is never reached leading to line `149` being executed again & again.

**Comment by krassowski** [comment]
> Though there is minor bug - The permission error above can occur on each key press in a notebook or on save when a server with `requires_documents_on_disk=False` is chosen.

Yes, if there is any server which requires document on disk this logic will be entered and errors will be raised repeatedly, but this should only happen for servers which do require a file to be on disk, as per this guard:

https://github.com/jupyter-lsp/jupyterlab-lsp/blob/b57638376714d2e48b4a8389570f983d0e3c1932/python_packages/jupyter_lsp/jupyter_lsp/virtual_documents_shadow.py#L126-L128

I think we can do three trials (to allow for transitive IO errors), and then just emit a warning informing about the shadow copy on disk system being unavailable and that users may want to adjust the path or if they maintain the server specification to try toggling `requires_documents_on_disk`. Does it sound good or do you have another suggested solution?

**Comment by nishikantparmariam** [comment]
> I think we can do three trials (to allow for transitive IO errors), and then just emit a warning informing about the shadow copy on disk system being unavailable and that users may want to adjust the path or if they maintain the server specification to try toggling requires_documents_on_disk.

This sounds good to me 👍🏽

**Comment by krassowski** [comment]
Closing as addressed by https://github.com/jupyter-lsp/jupyterlab-lsp/pull/936 (released in [v4.2.0](https://github.com/jupyter-lsp/jupyterlab-lsp/releases/tag/v4.2.0)).
